jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/bpmn2/BPMN2DataServiceImpl.java

```diff
@@ -131,9 +83,29 @@ private void validateNonEmptyDeploymentIdAndProcessIdAndTaskName(String deployme
             throw new IllegalStateException( msg.toString() );
         }
     }
+    
+    public void addProcessDefinition(String deploymentId, String processId, Object processDescriptor, KieContainer kieContainer) {
+        Map<String, ProcessDescriptor> definitions = null;
+        synchronized (definitionCache) {
+            definitions = definitionCache.get(deploymentId);
+            if (definitions == null) {
```

Comment:
I don't think I can use the putIfAbsent here as I need to check and only once initializes the map. So I need to check and get it anyway on subsequent calls. And the race condition is unlikely as that would mean same deployment has to be added at exact same time.